src/record-step-state-access.h

```diff
@@ -0,0 +1,840 @@
+        (void) mcycle_max;
+        return {mcycle, false};
+    }
+
+    uint64_t do_read_pma_istart(int i) const {
+        assert(i >= 0 && i < (int) PMA_MAX);
+        touch_page(shadow_pmas_get_pma_abs_addr(i));
+        const auto &pmas = m_m.get_pmas();
+        uint64_t istart = 0;
+        if (i >= 0 && i < static_cast<int>(pmas.size())) {
+            istart = pmas[i].get_istart();
+        }
+        return istart;
+    }
+
+    uint64_t do_read_pma_ilength(int i) const {
+        assert(i >= 0 && i < (int) PMA_MAX);
+        touch_page(shadow_pmas_get_pma_abs_addr(i));
+        const auto &pmas = m_m.get_pmas();
+        uint64_t ilength = 0;
+        if (i >= 0 && i < static_cast<int>(pmas.size())) {
+            ilength = pmas[i].get_ilength();
+        }
+        return ilength;
+    }
+
+    template <typename T>
+    void do_read_memory_word(uint64_t paddr, const unsigned char *hpage, uint64_t hoffset, T *pval) const {
+        (void) paddr;
+        touch_page(paddr);
+        *pval = cartesi::aliased_aligned_read<T>(hpage + hoffset);
+    }
+
+    template <typename T>
+    void do_write_memory_word(uint64_t paddr, unsigned char *hpage, uint64_t hoffset, T val) {
+        (void) paddr;
+        touch_page(paddr);
+        aliased_aligned_write(hpage + hoffset, val);
+    }
+
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    bool do_read_memory(uint64_t paddr, const unsigned char *data, uint64_t length) const {
+        (void) paddr;
+        (void) data;
+        (void) length;
+        throw std::runtime_error("Unexpected call to do_read_memory");
+    }
+
+    // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+    bool do_write_memory(uint64_t paddr, const unsigned char *data, uint64_t length) {
+
+        (void) paddr;
+        (void) data;
+        (void) length;
+        throw std::runtime_error("Unexpected call to do_write_memory");
+    }
+
+    template <typename T>
+    pma_entry &do_find_pma_entry(uint64_t paddr) {
+        int i = 0;
+        while (true) {
+            touch_page(shadow_pmas_get_pma_abs_addr(i));
+            auto &pma = m_m.get_state().pmas[i];
+            // The pmas array always contain a sentinel. It is an entry with
+            // zero length. If we hit it, return it
+            if (pma.get_length() == 0) {
+                return pma;
+            }
+            // Otherwise, if we found an entry where the access fits, return it
+            // Note the "strange" order of arithmetic operations.
+            // This is to ensure there is no overflow.
+            // Since we know paddr >= start, there is no chance of overflow
+            // in the first subtraction.
+            // Since length is at least 4096 (an entire page), there is no
+            // chance of overflow in the second subtraction.
+            if (paddr >= pma.get_start() && paddr - pma.get_start() <= pma.get_length() - sizeof(T)) {
+                touch_page(paddr);
```

Comment:
Why touch `paddr` itself here? Won't you touch it elsewhere afterwards?

Reply:
It is not needed. Good catch.